Implemented /join

Kaki-In opened the pull request:
# Implemented the /join command. 

## Change that have been added 

 - add a method with a `switch` syntax to improve the commands manager.
 - add the /join command - we can now join a room using /join #roomname:server.addr
 - improve the error message display (css)

Should fix [this issue (#792)](https://github.com/vector-im/hydrogen-web/issues/792).

Kaki-In committed:
Added the possibility to join a room using /join (also added the global commands uses, and some others commands like /shrug .)

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +197,80 @@ export class RoomViewModel extends ViewModel {
+    async _getMessageInformations (message) {
+        let msgtype = "m.text";
+        if (message.startsWith("/")) {
+            let command=message.substring(1).split(" ");
```
You can use [array destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#array_destructuring) here to make this a bit more readable:
```suggestion
            const [command, ...args] = message.substring(1).split(" ");
```
This way you don't have to use `command[0]` everywhere.

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +197,80 @@ export class RoomViewModel extends ViewModel {
+            let command=message.substring(1).split(" ");
+            switch (command[0]) {
+                case "me":
+                   message = message.substring(4).trim();
```
With the above suggestion, to get the rest of the message excluding the command, you can use:
```suggestion
                   message = args.join(" ").trim();
```
This is so you don't need to call `substring` with different argument for different commands.

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +197,80 @@ export class RoomViewModel extends ViewModel {
+               case "join":
+                   if (command.length==2) {
+                       let roomname = message.substring(5).trim();
+                       try {
```
A bit more simpler:
```suggestion
                   const [roomName] = args;
                   if (roomName) {
                       try {
```

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +197,80 @@ export class RoomViewModel extends ViewModel {
+                       try {
+                           await this._options.client.session.joinRoom(roomname);
+                       } catch (exc) {
+                           if (exc.statusCode??exc.status==400) {
```
We always add a space before and after operators:
```suggestion
                           if (exc.statusCode ?? exc.status === 400) {
```
Can you make this change throughout? It should be possible to setup your IDE/editor to do this automatically.

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +197,80 @@ export class RoomViewModel extends ViewModel {
+                       try {
+                           await this._options.client.session.joinRoom(roomname);
+                       } catch (exc) {
+                           if (exc.statusCode??exc.status==400) {
```
Also you probably meant to do `===` instead of `==`?

Kaki-In replied:
Why should we need to increase the verification level?

MidhunSureshR replied:
`===` is strict equality check, see https://stackoverflow.com/a/359509

Kaki-In replied:
Yes I know... But if the status is `"404"` instead of being `404` (and I am not sure that it is possible) what's the problem?

MidhunSureshR replied:
It's a code quality thing i.e if you have the choice of choosing between `===` or `==`, you should always choose `===`.

Kaki-In replied:
Ok, thanks

bwindels replied:
Basically, we always use === unless there is a good reason to use == (which hasn't happened yet I think?), so we are less likely to get it wrong when it does matter :)

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +197,80 @@ export class RoomViewModel extends ViewModel {
+                   if (command.length==2) {
+                       let roomname = message.substring(5).trim();
+                       try {
+                           await this._options.client.session.joinRoom(roomname);
```
After joining the room, we should switch over to the new room:
```suggestion
                           const internalId = await this._options.client.session.joinRoom(roomName);
                           this.navigation.push("room", internalId);
```

MidhunSureshR replied:
Also wanted to mention that we use camel casing for variables, so prefer `roomName` over `roomname`.

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +197,80 @@ export class RoomViewModel extends ViewModel {
+                        message = message.substring(1).trim();
+                        break;
+                    } else {
+                        this._sendError=new Error("no command name '"+command[0]+"'");
```
You can use a template string here if you want:
```suggestion
                        this._sendError=new Error(`no command named '${command[0]}'`);
```
We may also want to mention that you can use double slash `//` if you actually intend to send a message starting with `/`

MidhunSureshR replied:
It would also be nice to have a button to dismiss the error.

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +197,80 @@ export class RoomViewModel extends ViewModel {
+                   }
+           }
+       }
+       return [msgtype, message];
```
Do you think it would make more sense to return a object `{msgType, message}` rather than an array?

MidhunSureshR reviewed (commented):
Looking good, leaving some comments here:

Kaki-In committed:
Corrected some syntax dismiss

Kaki-In committed:
The error message can now be closed

Kaki-In committed:
Fixed priority operations when checking request status

Kaki-In committed:
Added the 403 status when joining an unknown room

Kaki-In committed:
Error of interpretation of the 403 status at the last update. Fixed

Kaki-In committed:
SyntaxError fixed

MidhunSureshR commented on src/platform/web/ui/session/room/RoomView.js:
```diff
@@ -46,7 +46,13 @@ export class RoomView extends TemplateView {
+						className: "RoomView_error_closerButton",
+						onClick: evt => vm.dismissError(evt)
+					}))
+                ]),
```
You can use one binding:
```js
t.div({className: "RoomView_error"}, [
	t.if(vm => vm.error, t => t.div(
		[
			t.p({}, vm => vm.error),
			t.button({ className: "RoomView_error_closerButton", onClick: evt => vm.dismissError(evt) })
		]
	)),
```

Kaki-In replied:
I need to modify the theme.css then.

Kaki-In committed:
The room is now joined after having actualised the rooms list, to avoid the synchronisations waits that can sometimes disable to enter the room (message "You're not into this room" or simply "You're not in this room yet. *Join the room*")

MidhunSureshR commented on src/matrix/Session.js:
```diff
@@ -948,6 +948,23 @@ export class Session {
         });
     }
+
+    waitForRoomFromSync(roomId) {
```
Can you integrate this with the `joinRoom` method so that it returns a Promise that eventually resolves to the joined room?

bwindels replied:
Given the onliner below, not sure it is worth it actually for now. It's true that the API is not super obvious, but I'd rather improve that separately.

MidhunSureshR replied:
True, resolving this in favor of https://github.com/vector-im/hydrogen-web/pull/809#discussion_r928833506

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +197,83 @@ export class RoomViewModel extends ViewModel {
-                    await replyingTo.reply(msgtype, message);
-                } else {
-                    await this._room.sendEvent("m.room.message", {msgtype, body: message});
+               let msgtype = messinfo.type;
```
Can you use `const` if you don't need to redefine the variables?

MidhunSureshR reviewed (commented):
Some more comments:

bwindels commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +197,83 @@ export class RoomViewModel extends ViewModel {
+                   msgtype = "m.emote";
+                   break;
+               case "join":
+                   if (args.length == 1) {
```
We strive to have methods not much longer than 15-20 lines (definitely with some exceptions); the code for handling the join command seems like a good candidate to extract in a separate method. Then you won't need the `               if (msgtype && message) {` on line 271 which on the face of it is a bit strange, why wouldn't you have a message if you just typed one?

Kaki-In replied:
if you type /join @roomid:site.location, you don't want to send an undefined object to everybody (it really does), so I had to precise that, if the message or the message type goes to be undefined, the message won't be sent

MidhunSureshR replied:
> why wouldn't you have a message if you just typed one?

There's nothing to send if an invalid command was entered, for eg: sending "/foobar" from the composer should not send any message.

Kaki-In replied:
For /join, there is also nothing to send, even if the room id was right

bwindels replied:
Yeah, you need the msgtype && message because some commands do want to send something, like tableflip. Missed that bit.

bwindels commented on src/matrix/Session.js:
```diff
@@ -948,6 +948,23 @@ export class Session {
+    waitForRoomFromSync(roomId) {
+        let resolve;
+        const promise = new Promise(r => { resolve = r; })
+        const subscription = {
```
This is a bit of a hacky way to do it, because you're not implementing the full IMapObserver contract, and it's a bit overkill anyway. You can use `session.observeRoomStatus` instead which is meant for this use case:
```js
await session.observeRoomStatus(roomId).waitFor(status === RoomStatus.Joined);
```

bwindels commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +197,83 @@ export class RoomViewModel extends ViewModel {
     
+    async _getMessageInformations (message) {
+        let msgtype = "m.text";
+        if (message.startsWith("/")) {
```
This method is more general than it needs to be. We don't do anything currently to messages apart from processing commands. You could put this `if` in the parent method and change this method to be `_processCommand` called from within that `if`.

bwindels commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +197,83 @@ export class RoomViewModel extends ViewModel {
+                   if (args.length == 1) {
+                       let roomName = args[0];
+                       try {
+                           const internalId = await this._options.client.session.joinRoom(roomName);
```
I'd just call this `roomId`, `internalId` suggests this is something else

Kaki-In committed:
- Application des différents commentaires du Pull Request (#809)
 - Correction des erreurs d'indentations.

Kaki-In committed:
Merge branch 'vector-im:master' into implement-join

Kaki-In committed:
Use "args.join" instead of "message.substring"
into RoomViewModel._processCommands

Kaki-In committed:
Fixed an error and improving css
If the /join command success, an error was thrown, because of a copy-pasted command not well integrated
The button of the error on "theme.css" contains now an unicode cross. The :after/:before cross was disformed when opening the room informations.

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +198,83 @@ export class RoomViewModel extends ViewModel {
+                    this._sendError = new Error("join syntax: /join <room-id>");
+                    this._timelineError = null;
+                    this.emitChange("error");
+                }
```
Can you extract this code into a separate method?

Kaki-In committed:
Separated the join instructions into a executeJoinCommand method

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +198,87 @@ export class RoomViewModel extends ViewModel {
         }
     }
     
+    async executeJoinCommand(args) {
```
It's a bit more cleaner to just pass the room id to this method.
so
```js
   case "join":
                const roomId = args[0];
                if(roomId) {
                   await this.executeJoinCommand(roomId);                
                }
                break;
```

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +198,87 @@ export class RoomViewModel extends ViewModel {
+            let roomName = args[0];
+            try {
+                const roomId = await this._options.client.session.joinRoom(roomName);
+                await (await this._options.client.session.observeRoomStatus(roomId)).waitFor(status => status === RoomStatus.Joined);
```
You don't need the double await here.

Kaki-In replied:
I need an await to get the `this._options.client.session.observeRoomStatus(roomId)` object (because he do have the .waitFor) and an await to wait the waitFor to be ended. 

I already tried without the second await, it throws a Type Error because `this._options.client.session.observeRoomStatus(roomId)` returns a Promise.

Kaki-In replied:
Is this code better? :  
```javascript
            const roomId = await this._options.client.session.joinRoom(roomName);
            const roomStatusObserver = await this._options.client.session.observeRoomStatus(roomId);
            await roomStatusObserver.waitFor(status => status === RoomStatus.Joined);
```

MidhunSureshR replied:
> Is this code better? :
> 
> ```js
>             const roomId = await this._options.client.session.joinRoom(roomName);
>             const roomStatusObserver = await this._options.client.session.observeRoomStatus(roomId);
>             await roomStatusObserver.waitFor(status => status === RoomStatus.Joined);
> ```


ah yes, you do need the double await since it's not a promise that resolves to another promise. This code looks good! :+1:

Kaki-In committed:
Separated the _processCommand and the joinRoom command
 - renamed executeJoinCommand as joinRoom;
 - separated the joinRoom process and the parse and result process

Kaki-In committed:
Some fixes :
 - fixed a pretty syntax miss (a !== b);
 - fixed a type error : replaced "msgtype" by "type" when instantied the "messinfo" variable;
 - some indentation fixes

Kaki-In committed:
Placed the join command outside of the processCommand method

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +198,100 @@ export class RoomViewModel extends ViewModel {
+        }
+    }
+    
+    async _processCommandJoin(roomName) {
```
I think this method can be merged with `joinRoom` method.

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +198,100 @@ export class RoomViewModel extends ViewModel {
+                msgtype = "m.emote";
+                break;
+            case "join":
+                if (args.length == 1) {
```
```suggestion
                if (args.length === 1) {
```

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +198,100 @@ export class RoomViewModel extends ViewModel {
+    } 
+
+    async _processCommand (message) {
+        let msgtype = undefined;
```
```suggestion
        let msgtype;
```

MidhunSureshR commented on src/domain/session/room/RoomViewModel.js:
```diff
@@ -197,18 +198,100 @@ export class RoomViewModel extends ViewModel {
+                this._sendError = new Error(`no command name "${commandName}". To send the message instead of executing, please type "/${message}"`);
+                this._timelineError = null;
+                this.emitChange("error");
+                msgtype = undefined;
```
`msgtype` would undefined anyway, so can remove this line.

MidhunSureshR commented on src/platform/web/ui/css/themes/element/theme.css:
```diff
@@ -521,6 +521,62 @@ a {
+}
+
+.RoomView_error:not(:empty) {
+    height : 40px;
```
Currently the errors are not fully shown in mobile devices.
```suggestion
    height : auto;
```

MidhunSureshR reviewed (approved):
Last couple of changes:

Kaki-In committed:
Finals checks about https://github.com/vector-im/hydrogen-web/pull/809#pullrequestreview-1053501341
 - joined the processJoinRoom and joinRoom methods;
 - fixed some precisions miss;
 - removed some useless code;
 - change the error message height from absolute (40px) to relative (auto)

Kaki-In committed:
Forgot one change

Kaki-In committed:
Review last checks

Kaki-In committed:
Fixed last check

Kaki-In committed:
Restored the common.js indentation

Kaki-In committed:
Pretty syntaxed the RoomViewModel
